New topic for pipeline chaining (#450)
harness/developer-hub: docs/platform/8_Pipelines/pipeline-chaining.md

---
title: Pipeline chaining in Harness
description: Chain your pipelines in Harness to create complex workflows.
sidebar_position: 12
---

:::note
Currently, this feature is behind the feature flag `PIPELINE_CHAINING`. Contact Harness Support to enable the feature.
:::

Pipeline chaining involves using the output of one pipeline as input for another. You can link multiple processing steps together and execute them sequentially, creating a more complex workflow.

> In Harness, you can create a pipeline chain by linking pipelines together using a pipeline trigger. A pipeline trigger is a specific event or condition that triggers the next pipeline in the chain. 

Author: Removed this line because the mention of pipeline trigger is not correct and might confuse customers.

> For example, you could have a pipeline that deploys an application to a test environment. This is followed by another pipeline that runs a set of integration tests, and finally another pipeline that deploys the application to production. Each pipeline in the chain is triggered by the completion of the previous pipeline. The output of each pipeline is passed on to the next pipeline in the chain. This way, you can automate the whole process and make sure that the application is deployed and tested properly before deploying to production.

Reviewer: ```suggestion
For example, you could have a pipeline that deploys an application to a test environment. This could be followed by a pipeline that runs a set of integration tests and, finally, a pipeline that deploys the application to production. Each pipeline in the chain is triggered by the completion of the previous pipeline. The output of each pipeline is passed to the next pipeline in the chain. This way, you can automate the whole process and make sure that the application is tested properly before deploying to production.
```
"Deployed and tested properly before deploying to production" reads odd because we say we deploy before we deploy. Is the current suggestion okay?)

description: Chain your pipelines in Harness to easily create complex workflows.
sidebar_position: 12
---

:::note
Currently, this feature is behind the feature flag `PIPELINE_CHAINING`. Contact Harness Support to enable the feature.
:::

Pipeline chaining involves using the output of one pipeline as input for another. You can link multiple processing steps together and execute them sequentially, creating a more complex workflow.

In Harness, you can create a pipeline chain by linking pipelines together using a pipeline trigger. A pipeline trigger is a specific event or condition that triggers the next pipeline in the chain. 

For example, you could have a pipeline that deploys an application to a test environment. This is followed by another pipeline that runs a set of integration tests, and finally another pipeline that deploys the application to production. Each pipeline in the chain is triggered by the completion of the previous pipeline. The output of each pipeline is passed on to the next pipeline in the chain. This way you can easily automate the whole process and make sure that the application is deployed and tested properly before deploying to production.

> ## Important

Author: This is the list of permissions required and the limitations this feature has. Putting them as a note or info would not be right in this context.

> ## Important

Reviewer: Okay, I agree.

- Pipeline stage does not support auto expressions in **Outputs**.

## Benefits of pipeline chaining

- Automation: Pipeline chaining allows you to automate the deployment process and make sure that the application is deployed and tested properly before deploying to production.
 
- Creation of complex workflows: It allows you to create complex workflows that involve multiple stages of deployment, testing, and verification.
 
- Error Handling: You can easily handle errors and rollback to previous stages in case of any issues.
 
- Visibility: It provides better visibility into the deployment process and makes it easier to identify and troubleshoot issues.
 
- Enhanced efficiency: By automating the deployment process, pipeline chaining can improve the efficiency and speed of deployments.
 
- Reusability: The pipelines can be reused across multiple applications and environments, reducing the need to recreate the same steps for each deployment.
 
- Better collaboration: By separating deployments into smaller, manageable stages, pipeline chaining allows different teams to work on different stages of the deployment process, improving collaboration and communication.
 

> - Better compliance: By automating the deployment process, pipeline chaining can help to ensure compliance with regulatory requirements and industry best practices.

Reviewer: Can we have a lead-in sentence and remove redundancies from the list? For example:
~~
With pipeline chaining, you can do the following:
- Automate the deployment process and make sure that the application is tested properly before deploying to production.
- Create complex workflows that involve multiple stages of deployment, testing, and verification.
- Handle errors and roll back to previous stages if issues occur.
(and so on)

~~
("Roll back" is two words when used as a verb.)
I would also question some of the items in the list. The first one, for example: isn't automation something we could do even before chained pipelines? It does not seem to say anything about chained pipelines. If automation is truly something introduced by this feature, we should elaborate.
The error handling one looks more like a useful feature we added and less like a direct benefit of being able to chain pipelines, which is what the title tells me to expect. 
I'm certain better collaboration and better compliance are direct benefits (as are the visibility and enhanced efficiency onces) but the statements will definitely benefit from your rewording, if you can look into it after sending it out the door (given you don't have time now).

Author: Done

## Chain pipelines 
To chain pipelines in Harness, perform the following steps: 

> 1. Create a [pipeline](../8_Pipelines/add-a-stage.md#step-1-create-a-pipeline) and configure it with the desired settings, such as triggers, environment variables, and deployment steps.
>    This is your parent pipeline.

Reviewer: ```suggestion
1. Create your parent [pipeline](../8_Pipelines/add-a-stage.md#step-1-create-a-pipeline) and configure it with the desired settings, such as triggers, environment variables, and deployment steps.
```
Until I reached the next line, I did not know which pipeline I was creating first... the parent or a child. Good to set context early.

Author: Done

2. Select **Add Stage** and then select **Pipeline**.
   
   ![](./static/pipeline-chain-option.png)

3. Select the pipeline from the list. You can select any pipeline across different orgs and projects corresponding to your access and permissions and chain it to your parent pipeline.
   
   For more information on access, see [Role-based access control](../4_Role-Based-Access-Control/1-rbac-in-harness.md)

   ![](./static/pipeline-chain-list.png)

4. Click **Apply Selected**.

> 5. You can use outputs from child pipelines in other stages in parent pipelines by clicking **Outputs**.

Reviewer: This sentence states a fact and needs a rewrite. Can we convert this to a step? Something like, "to do this, click **Outputs**.
Also, using the singular whenever appropriate makes for a clearer idea. Multiple plurals can lead to confusion and questions such as "can I use the output of one child pipeline in multiple other stages in the parent?" or "can I use the output of multiple child pipelines in a subsequent stage?" Consider this example: "To use the output of the child pipeline in subsequent stages of the parent pipeline, click ..." (Then you are referring to this specific child pipeline you are documenting and using the singular.)

Author: Done.

   The variables you select in the output tab can be referred to at a later stage.

>    For example, you can use the build image in the subsequent deployment stage.

Reviewer: ```suggestion
   For example, you can use the build image in a subsequent deployment stage.
```

   
   :::note
   The outputs tab does not support auto suggestions. To use the child execution expression in the parent pipeline, you must manually copy it.
   :::
   

6. Repeat this process for any additional pipelines that you want to chain.

> 7. In the final pipeline, you can add a **Deploy** step to deploy the application to the desired environment.

Reviewer: final child pipeline?

Author: Rephrased to "final chained pipeline".

>    You can also add "Rollback" step in case of any failure.

Reviewer: ```suggestion
   You can also add a "Rollback" step if your deployment fails.
```
(Is my correction accurate?)

Author: This statement needs to be removed. Rollback is not supported yet.

>    Once you have completed chaining the pipelines, you can run the parent pipeline to start the process.

Reviewer: ```suggestion
   After you have chained the pipelines, you can run the parent pipeline.
```

 
- Creation of complex workflows: It allows you to create complex workflows that involve multiple stages of deployment, testing, and verification.
 
- Error Handling: You can easily handle errors and rollback to previous stages in case of any issues.
 
- Visibility: It provides better visibility into the deployment process and makes it easier to identify and troubleshoot issues.
 
- Enhanced efficiency: By automating the deployment process, pipeline chaining can improve the efficiency and speed of deployments.
 
- Reusability: The pipelines can be reused across multiple applications and environments, reducing the need to recreate the same steps for each deployment.
 
- Better collaboration: By separating deployments into smaller, manageable stages, pipeline chaining allows different teams to work on different stages of the deployment process, improving collaboration and communication.
 
- Better compliance: By automating the deployment process, pipeline chaining can help to ensure compliance with regulatory requirements and industry best practices.

> - Fast rollback: It makes it easy to roll back to previous stages in case of any issues.

Reviewer 2: Rollback is not supported in prod as of now. Not it has been supported in pipeline chaining
